Add a new version of csvfind

alan910127 opened the pull request:
Try to improve the Rust version:
- Reduce the amount of heap allocations
- Use more iterators
- Add a `contains_ignore_ascii_case` function
- Make the `header_name` to be an `Option` like you mentioned in the video

It looks like the functionality is working fine, but I haven't tested the performance

alan910127 committed:
Add a new version of csvfind

Try to improve the Rust version:
- Reduce the amount of heap allocations
- Use more iterators
- Add a `contains_ignore_ascii_case` function
- Make the `header_name` to be an `Option` like you mentioned in the video

It looks like the functionality is working fine, but I haven't tested the performance

unconv commented:
Nice!! It's actually faster than the C version, too!

```
0.788373s: php csvfind.php output.csv winkpad
0.533784s: ./c-csvfind-test output.csv winkpad
0.454778s: ./rust-csvfind-test output.csv winkpad
```

Although sometimes the C version is faster. Of course my test is not the most accurate...
```
0.841662s: php csvfind.php output.csv winkpad
0.513793s: ./c-csvfind-test output.csv winkpad
0.545371s: ./rust-csvfind-test output.csv winkpad
```

Your version doesn't print the headers, but adding `println!( "{}", headers.join(",") );` before the loop fixes that (I couldn't figure out how to print the raw first line after `let headers` calls `lines.next()` already). That also doesn't have an effect on the performance since it's outside of the loop. Interestingly, when I tried to naively set `let mut first_line = true;` and then set it to false inside the loop, this doubled the execution time. So with this knowledge I can now probably speed up the C implementation even more :)

Thanks!

alan910127 commented:
You can use the `headers` before split, print it, and use shadowing to transform it into a `Vec` later. 

```rust
let headers = lines
    .next()
    .expect("Headers should be presented in the CSV file")?;

println!("{}", headers);

let headers = headers
    .split(',')
    .map(|s| s.trim().to_string())
    .collect::<Vec<_>>();
```

In this way, you don't even need to "own" the headers by keeping the ownership of the raw header line in a designated variable. Perhaps this can be a little bit more performant than the previous version I wrote.
(But I think it's not that important since it only contains 5 strings)

```rust
let raw_headers = lines
    .next()
    .expect("Headers should be presented in the CSV file")?;

println!("{}", raw_headers);

let headers = raw_headers
    .split(',')
    .map(|s| s.trim())
    .collect::<Vec<_>>();
```
